### PR TITLE
Added unassigned_role and tactic role request edge case

### PR DESCRIPTION
**Pull request by PotentialError**
## Description
Added an unassigned_role and logic to handle unassigned robots and tactics which have too many roles relative to available robots.
**FOR FUTURE TACTIC WRITERS:** After creating new role requests, wait until len(self.assigned_roles) == # of needed roles to continue through your tactic!

## Steps to test
1. Run `ros2 launch rj_robocup sim.launch.py`with a play and tactics that will have robots unassigned
2. Run the play in the simulator

Expected result: In the command prompt, the message "UnassignedRole created for: {robot info}" should be printed as many times as there are unassigned robots

### CODE REVIEW

**Review by kfu02 — APPROVED** [review]
Looks good, just fix the style errors

**Review by Yudai-8-om — COMMENTED** [review]
[gameplay_node-11]     load_entry_point('rj-gameplay', 'console_scripts', 'gameplay_node')()
[gameplay_node-11]   File "/home/vagrant/RoboJackets/robocup-software/rj_gameplay/rj_gameplay/gameplay_node.py", line 464, in main
[gameplay_node-11]     rclpy.spin(gameplay)
[gameplay_node-11]   File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/__init__.py", line 191, in spin
[gameplay_node-11]     executor.spin_once()
[gameplay_node-11]   File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 714, in spin_once
[gameplay_node-11]     raise handler.exception()
[gameplay_node-11]   File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/task.py", line 239, in __call__
[gameplay_node-11]     self._handler.send(None)
[gameplay_node-11]   File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 429, in handler
[gameplay_node-11]     await call_coroutine(entity, arg)
[gameplay_node-11]   File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 343, in _execute_timer
[gameplay_node-11]     await await_or_execute(tmr.callback)
[gameplay_node-11]   File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 118, in await_or_execute
[gameplay_node-11]     return callback(*args)
[gameplay_node-11]   File "/home/vagrant/RoboJackets/robocup-software/rj_gameplay/rj_gameplay/gameplay_node.py", line 222, in gameplay_tick
[gameplay_node-11]     intents = self.test_play.tick(self.world_state)
[gameplay_node-11]   File "/home/vagrant/RoboJackets/robocup-software/rj_gameplay/rj_gameplay/play/basic122.py", line 66, in tick
[gameplay_node-11]     return self.get_robot_intents(world_state)
[gameplay_node-11]   File "/home/vagrant/RoboJackets/robocup-software/rj_gameplay/stp/play/__init__.py", line 127, in get_robot_intents
[gameplay_node-11]     role_robot_intents = tactic.tick(world_state)
[gameplay_node-11]   File "/home/vagrant/RoboJackets/robocup-software/rj_gameplay/rj_gameplay/tactic/basic_seek.py", line 29, in tick
[gameplay_node-11]     self.init_roles(world_state)
[gameplay_node-11]   File "/home/vagrant/RoboJackets/robocup-software/rj_gameplay/rj_gameplay/tactic/basic_seek.py", line 43, in init_roles
[gameplay_node-11]     robot = self.assigned_robots[0]
[gameplay_node-11] IndexError: list index out of range
[ERROR] [gameplay_node-11]: process has died [pid 15647, exit code 1, cmd '/home/vagrant/RoboJackets/robocup-software/install/lib/rj_robocup/gameplay_node --ros-args --params-file /home/vagrant/RoboJackets/robocup-software/install/share/rj_robocup/config/sim.yaml'].

**Review by kfu02 — CHANGES_REQUESTED** [review]
When running basic122 (the default offense play) I'm getting a consistent crash on the pass tactic when it tries to assign roles: 
```
[gameplay_node-11]   File "/home/robojackets/robocup-software/rj_gameplay/rj_gameplay/gameplay_node.py", line 222, in gameplay_tick
[gameplay_node-11]     intents = self.test_play.tick(self.world_state)
[gameplay_node-11]   File "/home/robojackets/robocup-software/rj_gameplay/rj_gameplay/play/basic122.py", line 114, in tick
[gameplay_node-11]     return self.get_robot_intents(world_state)
[gameplay_node-11]   File "/home/robojackets/robocup-software/rj_gameplay/stp/play/__init__.py", line 130, in get_robot_intents
[gameplay_node-11]     role_robot_intents = tactic.tick(world_state)
[gameplay_node-11]   File "/home/robojackets/robocup-software/rj_gameplay/rj_gameplay/tactic/pass_tactic.py", line 119, in tick
[gameplay_node-11]     receiver_role = self.assigned_roles[1]
[gameplay_node-11] IndexError: list index out of range
```
Not sure why, let me know if you can't reproduce

File: rj_gameplay/rj_gameplay/role/unassigned_role.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import stp
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/role/unassigned_role.py]: Sadly, I think importing stp doesn't force the imports of all the subpackages (stp.role, stp.rc, etc.). You have to manually do this to fix the mypy errors below.

File: rj_gameplay/rj_gameplay/tactic/pass_tactic.py [hunk 2]
```diff
@@ -108,7 +108,8 @@ def tick(
 
         elif self._state == State.INIT_EXECUTE_PASS:
             # one tick delay for play role assignment
-            self._state = State.EXECUTE_PASS
+            if len(self.assigned_roles) == 2:
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/tactic/pass_tactic.py:111]: Nice fix, can you mention this in the description of your PR as a useful thing for future tactic writers?

File: rj_gameplay/stp/play/__init__.py [hunk 3]
```diff
@@ -87,41 +90,45 @@ def assign_roles(
 
         used_robots = set()
         for tactic in self.prioritized_tactics:
-            # TODO: handle if tactic requests more roles than exist
-            #       by not assigning tactic at all!
-            #       and give some default behavior (later)
             robots_for_tactic = []
```
Comment (kfu02) [inline rj_gameplay/stp/play/__init__.py:96]: Could you leave a comment explaining what robots_for_tactic represents/how the behavior has changed? I believe currently that it's something like "check if all the robots a tactic requests can be given to it: if yes, assign them all, if no, don't assign any of them"

File: rj_gameplay/stp/play/__init__.py [hunk 4]
```diff
@@ -87,41 +90,45 @@ def assign_roles(
...
                         cheapest_robot = robot
 
                 if cheapest_robot is None:
-                    # TODO: consider looping until success?
-                    raise RoleAssignFailure(
-                        f"RoleRequest ({role}, {cost_fn}) was not assigned"
-                    )
-
-                used_robots.add(cheapest_robot)
+                    print(f"RoleRequest ({role}, {cost_fn}) was not assigned")
```
Comment (kfu02) [inline rj_gameplay/stp/play/__init__.py:117]: Any reason you got rid of the RoleAssignFailure?

File: rj_gameplay/stp/utils/world_state_converter.py [hunk 5]
```diff
@@ -265,10 +265,10 @@ def robot_creator(
     else:
         is_ours = True
         robot_id = robot_state.id
-        ball_sense = Optional[robot_status.has_ball_sense]
-        kicker_charged = Optional[robot_status.kicker_charged]
-        kicker_healthy = Optional[robot_status.kicker_healthy]
-        lethal_fault = Optional[robot_status.lethal_fault]
+        ball_sense = robot_status.has_ball_sense
```
Comment (kfu02) [inline rj_gameplay/stp/utils/world_state_converter.py:268]: ignore these mypy errors for now

File: rj_gameplay/stp/play/__init__.py [hunk 6]
```diff
@@ -87,41 +90,45 @@ def assign_roles(
...
 
-        for tactic in self.prioritized_tactics:
-            tactic.init_roles(world_state)
+            if robots_for_tactic is not None:
+                used_robots.update(robots_for_tactic)
+                tactic.set_assigned_robots(robots_for_tactic)
+                self.approved_prioritized_tactics.append(tactic)
+                tactic.init_roles(world_state)
+            else:
+                print("Tactic denied")
```
Comment (kfu02) [inline rj_gameplay/stp/play/__init__.py]: Would be helpful if this printed something like "Tactic <name> denied: requested <x> roles, but only <y> robots available" or something of that nature

File: rj_gameplay/rj_gameplay/role/unassigned_role.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+from rj_msgs.msg import RobotIntent
+
+
+class UnassignedRole(stp.role.Role):
+    """Role given to help role assignment correctly handle an unassigned robot"""
+
+    def __init__(self, robot: stp.rc.Robot) -> None:
+        super().__init__(robot)
+        print(f"UnassignedRole created for {robot}!")
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/role/unassigned_role.py]: for this print, can you make it only print if the robot.visible flag is True? that way we avoid a lot of unnecessary prints that clutter the output

**Review by kfu02 — CHANGES_REQUESTED** [review]
Hey, left a few comments. The new Unassigned role looks good, but I have some concerns with the debug output it's giving now. Also, can you merge ros2 into your branch? There's been some changes recently and I suspect you'll have merge conflicts.

File: rj_gameplay/stp/utils/world_state_converter.py (hunk 5 above)
Comment (PotentialError) [reply rj_gameplay/stp/utils/world_state_converter.py:268]: Simulator on my end does not work without this change

File: rj_gameplay/stp/play/__init__.py (hunk 4 above)
Comment (PotentialError) [reply rj_gameplay/stp/play/__init__.py:117]: If I reintroduced it the program would crash but I assumed my code was supposed to allow play to continue and be able to handle too many role requests without throwing errors

File: rj_gameplay/rj_gameplay/role/unassigned_role.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+class UnassignedRole(stp.role.Role):
+    """Role given to help role assignment correctly handle an unassigned robot"""
+
+    def __init__(self, robot: stp.rc.Robot) -> None:
+        super().__init__(robot)
+        if robot.visible:
+            print(f"UnassignedRole created for {robot}!")
+
+    def tick(self, world_state: stp.rc.WorldState) -> RobotIntent:
+        """Send debug message to prove it is setting UnassignedRole"""
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/role/unassigned_role.py]: nit: this comment should be in the init

File: rj_gameplay/stp/play/__init__.py (hunk 4 above)
Comment (kfu02) [reply rj_gameplay/stp/play/__init__.py:117]: gotcha

File: rj_gameplay/stp/play/__init__.py [hunk 9]
```diff
@@ -81,47 +84,67 @@ def assign_roles(
...
+                self.approved_prioritized_tactics.append(tactic)
+                tactic.init_roles(world_state)
+            else:
+                numRobotsAvailable = (
+                    len(world_state.our_robots)
+                    - len(used_robots)
+                    - numOfInvisibleRobots
+                )
+                print(
+                    f"Tactic denied: {len(tactic.role_requests)} requested roles, but only {numRobotsAvailable} robots available"
```
Comment (kfu02) [inline rj_gameplay/stp/play/__init__.py]: nit: add tactic name to the debug output
```suggestion
                    f"Tactic {tactic} denied: {len(tactic.role_requests)} requested roles, but only {numRobotsAvailable} robots available"
```

**Review by kfu02 — APPROVED** [review]
Couple nits. Overall looks good and works as described. Nice job!

**Review by Yudai-8-om — APPROVED** [review]
LGTM, Great job!

**Pull request merged**